### PR TITLE
Dependency injection

### PR DESCRIPTION
**Pull request by Devilhar**

### CODE REVIEW

File: Game/Character/CharacterFunctionality.h [hunk 1]
```diff
@@ -29,6 +31,7 @@ class CharacterFunctionality
 	: public Functionality<
 		CharacterFunctionality,
 		Policy::Set<
+			Policy::Add<InputSystem, Policy::EPermission::Write>,
```
Comment (nicolasgustafsson) [inline Game/Character/CharacterFunctionality.h:34]: Why does CharacterFunctionality need InputSystem access? I might be missing something

File: Game/Designers/BuffDesigner/BuffDesigner.cpp [hunk 2]
```diff
@@ -43,24 +44,24 @@ BuffBlueprint& BuffDesigner::DesignBuff(SBuffRequirements aBuffRequirements)
...
 {
-	return std::powf(SystemPtr<Randomizer>()->GetRandomNumber(1.f, 1000.f) / 1000.f, 2.f) * 10.f;
+	return std::powf(Get<Randomizer>().GetRandomNumber(1.f, 1000.f) / 1000.f, 2.f) * 10.f;
 }
 
-void BuffDesigner::MakeBuffBetter(const f32 aHowMuch, SBuffDesign& aBuffDesign) const
+void BuffDesigner::MakeBuffBetter(const f32 aHowMuch, SBuffDesign& aBuffDesign)
 {
-	switch (SystemPtr<Randomizer>()->GetRandomNumber<i32>(0, 1))
+	switch (Get<Randomizer>().GetRandomNumber<i32>(0, 1))
```
Comment (nicolasgustafsson) [inline Game/Designers/BuffDesigner/BuffDesigner.cpp:64]: Might be worth thinking about having GetRandomNumber be const and making the seed mutable; getting a random number _feels_ like a pure operation

File: Game/Designers/LevelDesigner/LevelDesigner.h [hunk 3]
```diff
@@ -46,13 +56,25 @@ struct SRoom
...
+		Policy::Set<
+			Policy::Add<Randomizer, Policy::EPermission::Write>,
+			Policy::Add<BuffDesigner, Policy::EPermission::Write>,
+			Policy::Add<FunctionalitySystemDelegate<LevelPortalFunctionality>, Policy::EPermission::Write>,
+			Policy::Add<FunctionalitySystemDelegate<SpinnerFunctionality>, Policy::EPermission::Write>,
+			Policy::Add<FunctionalitySystemDelegate<TransformFunctionality>, Policy::EPermission::Write>,
+			Policy::Add<FunctionalitySystemDelegate<SpriteRenderingFunctionality>, Policy::EPermission::Write>,
+			Policy::Add<FunctionalitySystemDelegate<BuffGiverFunctionality>, Policy::EPermission::Write>,
+			Policy::Add<FunctionalitySystemDelegate<EnemyControllerFunctionality>, Policy::EPermission::Write>,
+			Policy::Add<FunctionalitySystemDelegate<StaticGeometryFunctionality>, Policy::EPermission::Write>>>
```
Comment (nicolasgustafsson) [inline Game/Designers/LevelDesigner/LevelDesigner.h:69]: Having delegates for this is clever ngl

File: Game/HitShapes/HitShapeSpawnerFunctionality.h [hunk 4]
```diff
@@ -14,19 +17,19 @@ enum class EDirection
...
-	Policy::Add<FactionFunctionality, Policy::EPermission::Read>,
-	Policy::Add<SFactionComponent, Policy::EPermission::Write>,
-	Policy::Add<SLevelDenizenComponent, Policy::EPermission::Write>>>
+		Policy::Set<
+			Policy::Add<Randomizer, Policy::EPermission::Write>,
+			Policy::Add<FunctionalitySystemDelegate<HitShapeFunctionality>, Policy::EPermission::Write>,
+			Policy::Add<TransformFunctionality, Policy::EPermission::Write>,
+			Policy::Add<FactionFunctionality, Policy::EPermission::Read>,
+			Policy::Add<SFactionComponent, Policy::EPermission::Write>,
+			Policy::Add<SLevelDenizenComponent, Policy::EPermission::Write>>>
```
Comment (nicolasgustafsson) [inline Game/HitShapes/HitShapeSpawnerFunctionality.h:26]: Feels like we could simplify Policy::Add<TPolicyType, Policy::EPermission::Write> to something like Write<TPolicyType> to reduce boilerplate

File: WonderMake/Functionalities/FunctionalitySystem.h [hunk 5]
```diff
@@ -1,45 +1,138 @@
...
+	TPolicy::Permission>;
+
+template<typename... TFunctionalityPolicies>
+using CreatePolicySet =
+Policy::Set<
+	ConvertPolicy<TFunctionalityPolicies>...>;
+
+template<typename TPolicySet>
+using ConvertPolicySet =
+typename TPolicySet::template ExtractPolicies<CreatePolicySet>;
```
Comment (nicolasgustafsson) [inline WonderMake/Functionalities/FunctionalitySystem.h:47]: breh

File: WonderMake/Program/Program.h [hunk 6]
```diff
@@ -30,22 +30,19 @@ class Program
...
-	SystemPtr<TimeKeeper> myTimeKeeperPtr;
 	SystemPtr<InputSystem> myInputSystem;
-	SystemPtr<DebugSettingsSystem> myDebugSettingsPtr;
 	SystemPtr<AudioManager> myAudioManagerPtr;
 	SystemPtr<FileWatcher> myFileWatcher;
+	SystemPtr<TimeKeeper> myTimeKeeperPtr;
+	SystemPtr<EngineUniformBuffer> myEngineUniformBufferPtr;
+	SystemPtr<Window> myWindowPtr;
+	SystemPtr<Camera> myCameraPtr;
+	SystemPtr<Renderer> myRendererPtr;
```
Comment (nicolasgustafsson) [inline WonderMake/Program/Program.h:45]: Assuming these will disappear later on?

File: WonderMake/System/System.h [hunk 7]
```diff
@@ -1,19 +1,69 @@
...
+	inline constexpr System(Dependencies&& aDependencies) noexcept;
+
+protected:
+	constexpr System() noexcept = default;
+	
+	template<
+		typename TDependency,
+		typename = std::enable_if_t<
+			TPolicySet::template HasPolicy_v<TDependency, Policy::EPermission::Write>
+			|| TPolicySet::template HasPolicy_v<TDependency, Policy::EPermission::Unrestricted>>>
```
Comment (nicolasgustafsson) [inline WonderMake/System/System.h:44]: requires? Or did it not work properly here?

File: Game/Character/CharacterFunctionality.h (hunk 1 above)
Comment (Devilhar) [reply Game/Character/CharacterFunctionality.h:34]: Will remove

File: WonderMake/Functionalities/FunctionalitySystem.h (hunk 5 above)
Comment (Devilhar) [reply WonderMake/Functionalities/FunctionalitySystem.h:47]: breh

File: WonderMake/Program/Program.h (hunk 6 above)
Comment (Devilhar) [reply WonderMake/Program/Program.h:45]: Correct. To remove them I'd need to make some refactoring which is beside what this pull request is intended to do.

File: WonderMake/System/System.h (hunk 7 above)
Comment (Devilhar) [reply WonderMake/System/System.h:44]: I do not remember, will have to check it out

File: Game/Character/CharacterFunctionality.h (hunk 1 above)
Comment (Devilhar) [reply Game/Character/CharacterFunctionality.h:34]: Never mind, it's used by WeaponSwingHitShapeAction.

File: Game/HitShapes/HitShapeSpawnerFunctionality.h (hunk 4 above)
Comment (Devilhar) [reply Game/HitShapes/HitShapeSpawnerFunctionality.h:26]: Agreed, will add them to the global scope and rename them.

**Pull request merged**